src/main/java/com/spotify/futures/CompletableFutures.java

```diff
@@ -83,6 +113,26 @@ private CompletableFutures() {
         });
   }
 
+  private static <T> void setupFailFast(
+      FailFast failFast, CompletableFuture<? extends T>[] all, CompletableFuture<Void> allOf) {
+    ExecutionMetadata executionMeta = new ExecutionMetadata(all);
```

Comment:
Do we really need the `ExecutionMetadata` type? Seems like the information is not used in the `FailFastWithThrowable ` implementation, and I'm not sure it's needed otherwise either.

Reply:
You are correct. I planed the proposal to provide the base functionality to fail fast but also to allow the extension of behavior in a safe manner. I added the execution metadata because I foresee that you cannot have a "solution to fit all" when it comes to cancel the futures or failing fast. 

Consider 8 tasks being joined. If 7 of them completed normally and the last one executing fails, would completing exceptionally be the right course of action? `FailFastWithThrowable` represents the simplest approach to fail fast behavior. Other implementations of `FailFast` might take perceptual thresholds before shot-circuiting executions and returning.

Reply:
Right, I see what you mean. Given that `allAsList` and `joinList` will always fail if any of their inputs fail (unlike `successfulAsList`), I think we'd need to find a different solution for this. Maybe that can happen before the `joinList`?
```java
collection.stream()
    .map(this::someAsyncFunction)
    .map(stage -> stage.handle(...)) // <-- fallback to a successful stage in certain cases
    .collect(CompletableFutures.joinList())
    .thenApply(this::consumeList)
```

Reminds me a bit of [Futures.catching(...) from Guava](https://github.com/google/guava/blob/master/guava/src/com/google/common/util/concurrent/Futures.java#L294). Maybe we could add something similar? Would suggest a separate PR for that though.